vignettes/test-fixtures.Rmd

```diff
@@ -0,0 +1,317 @@
+
+```{r}
+message2 <- function(...) {
+  if (!isTRUE(getOption("verbose"))) {
+    return()
+  }
+  message(...)
+}
+```
+
+In some cases, it's possible to make the global state an explicit argument to the function. For example, we could refactor `message2()` to make the verbosity an explicit argument:
+
+```{r}
+message3 <- function(..., verbose = getOption("verbose")) {
+  if (!isTRUE(verbose)) {
+    return()
+  }
+  message(...)
+}
+```
+
+Making external state explicit is often worthwhile, because it makes it more clear exactly what inputs determine the outputs of your function. But it's simply not possible in many cases. That's where test fixtures come in: they allow you to temporarily change global state in order to test your function. Test fixture is a pre-existing term in the software engineering world (and beyond):
+
+> A test fixture is something used to consistently test some item, device, or piece of software.
+>
+> --- [Wikipedia](https://en.wikipedia.org/wiki/Test_fixture)
+
+A **test fixture** is just a `local_` function that you use to change state in such a way that you can reach inside and test parts of your code that would otherwise be challenging. For example, here's how you could use `withr::local_options()` as a test fixture to test `message2()`:
+
+```{r}
+test_that("message2() output depends on verbose option", {
+  withr::local_options(list(verbose = TRUE))
+  expect_message(message2("Hi!"))
+  
+  withr::local_options(list(verbose = FALSE))
+  expect_message(message2("Hi!"), NA)
+})
+```
+
+### Case study: usethis
+
+One place that we use test fixtures extensively is in the usethis package ([usethis.r-lib.org](https://usethis.r-lib.org)), which provides functions for looking after the files and folders in R projects, especially packages. Many of these functions only make sense in the context of a package, which means to test them, we also have to be working inside an R package. We need a way to quickly spin up a minimal package in a temporary directory, then test some functions against it, then destroy it.
+
+To solve this problem we create a test fixture, which we place in `R/test-helpers.R` so that's it's available for both testing and interactive experimentation:
+
+```{r}
+create_local_package <- function(dir = file_temp(), env = parent.frame()) {
+  old_project <- proj_get_()
+  
+  # create new folder and package
+  create_package(dir, open = FALSE) # A
+  withr::defer(fs::dir_delete(dir), envir = env) # -A
+  
+  # change working directory
+  setwd(dir) # B
+  withr::defer(setwd(old_project), envir = env) # -B
+  
+  # switch to new usethis project
+  proj_set(dir) # C
+  withr::defer(proj_set(old_project, force = TRUE), envir = env) # -C
+  
+  dir
+}
+```
+
+(If we wrote this today, would call it `local_create_package()` to emphasise the local-ness of it.)
```

Comment:
I'm happy to rename it in usethis. I don't have any attachment to the name.

Reply:
Ok, renamed. You can change in usethis at your leisure; I don't think there's any rush.